Refactor dist tarballs generation

pietroalbini opened the pull request:
Before this PR, each tarball we ship as part of a release was generated by manually creating the directory structure and invoking `rust-installer generate`. This means each tarball was slightly different, adding new ones meant copy-pasting the code generating another tarball and removing the useless parts, and more importantly refactoring how tarballs are generated is extremely time-consuming.

This PR introduces a new abstraction in rustbuild, `Tarball`. The `Tarball` struct provides a trivial API to generate simple tarballs, and can get out of the way when more complex tarballs have to generate. For example, the whole code to generate the `build-manifest` tarball is now the following:

```rust
let tarball = Tarball::new(builder, "build-manifest", &self.target.triple);
tarball.add_file(&build_manifest, "bin", 0o755);
tarball.generate()
```

One notable change between the old tarballs and the new ones is that the "overlay" (README.md, COPYRIGHT, LICENSE-APACHE and LICENSE-MIT) is now available in every produced tarball, while before each tarball inconsistently had or didn't have those files. Tarballs that need a different overlay have a way to change which files to include (with the `set_overlay` method):

```rust
let mut tarball = Tarball::new(builder, "rustfmt", &target.triple);
tarball.set_overlay(OverlayKind::Rustfmt);
tarball.add_file(rustfmt, "bin", 0o755);
tarball.add_file(cargofmt, "bin", 0o755);
tarball.add_legal_and_readme_to("share/doc/rustfmt");
Some(tarball.generate())
```

The PR should be reviewed commit-by-commit, as each commit migrated a separate tarball to use `Tarball`. During development i made sure every tarball can still be generated, and for the most compex tarballs I manually ensured the list of files between the old and new tarballs did not have unexpected changes.

r? @Mark-Simulacrum

Keruspe commented on src/bootstrap/tarball.rs:
```diff
@@ -206,9 +205,40 @@ impl<'a> Tarball<'a> {
                 .arg(input_tarballs)
                 .arg("--non-installed-overlay")
                 .arg(&this.overlay_dir);
```
I think you can drop the two args above?

pietroalbini replied:
Fixed!

Mark-Simulacrum commented:
I'm probably not going to get a chance to review this until the weekend at the earliest -- is this blocking some work? If so I could potentially fit it in but it'd be a bit rough.

pietroalbini commented:
It's not blocking any work, the next task for me on this project is to change `rust-installer`, and that can happen in parallel of this PR. Also I'm probably as busy as you :)

Mark-Simulacrum commented on src/bootstrap/tarball.rs:
```diff
@@ -48,6 +49,11 @@ impl OverlayKind {
+            OverlayKind::RustAnalyzer => &[
+                "src/tools/rust-analyzer/README.md",
+                "src/tools/rls/LICENSE-APACHE",
+                "src/tools/rls/LICENSE-MIT",
```
uh, typo?

pietroalbini replied:
Fixed!

Mark-Simulacrum commented:
This looks amazing to me -- I noted one typo, and I'm sure I might've missed others, but I don't think that should block (we can patch things up if we notice them later). It seems good to get this landed quickly so we're not landing it around the release (so we can fix things on master without worrying about backports).

r=me with typo fixed -- up to you on @Keruspe's comment.

pietroalbini commented:
@bors r=Mark-Simulacrum

bors commented:
:pushpin: Commit 272838e3f60487a40e25363e6f681dfd011234dc has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum 272838e3f60487a40e25363e6f681dfd011234dc -->
<!-- homu: {"type":"Approved","sha":"272838e3f60487a40e25363e6f681dfd011234dc","approver":"Mark-Simulacrum"} -->

bors commented:
:hourglass: Testing commit 272838e3f60487a40e25363e6f681dfd011234dc with merge de7ce5c3f8869f3288d1fb9480fdcd985362f15a...
<!-- homu: {"type":"BuildStarted","head_sha":"272838e3f60487a40e25363e6f681dfd011234dc","merge_sha":"de7ce5c3f8869f3288d1fb9480fdcd985362f15a"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1550190296)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1550190296","builder_name":"checks-actions"} -->

pietroalbini commented:
@bors try

bors commented:
:hourglass: Trying commit 272838e3f60487a40e25363e6f681dfd011234dc with merge 80acf4fa64fccb397fb4cec1d663dd473f492507...
<!-- homu: {"type":"TryBuildStarted","head_sha":"272838e3f60487a40e25363e6f681dfd011234dc","merge_sha":"80acf4fa64fccb397fb4cec1d663dd473f492507"} -->

Mark-Simulacrum commented:
FWIW I think git absorb or similar should make it easy to avoid the "fix review comments" commit, but since this would still have 21 commits - which seems right - I am not going to block on that.

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1550473112)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1550473112","builder_name":"checks-actions"} -->

pietroalbini commented:
Ok should've fixed the CI failure (I also amended the commit history).

@bors try

bors commented:
:hourglass: Trying commit 6a66aa4a263a3cd7586b723ec589c2f739e233ef with merge bf52e3fae8d78a18529a0c726ce78238217c20e8...
<!-- homu: {"type":"TryBuildStarted","head_sha":"6a66aa4a263a3cd7586b723ec589c2f739e233ef","merge_sha":"bf52e3fae8d78a18529a0c726ce78238217c20e8"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1550846223)
Build commit: bf52e3fae8d78a18529a0c726ce78238217c20e8 (`bf52e3fae8d78a18529a0c726ce78238217c20e8`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1550846223"},"merge_sha":"bf52e3fae8d78a18529a0c726ce78238217c20e8"} -->

pietroalbini commented:
@bors r=Mark-Simulacrum

bors commented:
:pushpin: Commit 6a66aa4a263a3cd7586b723ec589c2f739e233ef has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum 6a66aa4a263a3cd7586b723ec589c2f739e233ef -->
<!-- homu: {"type":"Approved","sha":"6a66aa4a263a3cd7586b723ec589c2f739e233ef","approver":"Mark-Simulacrum"} -->

pietroalbini commented:
@bors rollup=iffy

bors commented:
:hourglass: Testing commit 6a66aa4a263a3cd7586b723ec589c2f739e233ef with merge 714bbd995c711cb7242dbf66cb6aa2c3a238cfb7...
<!-- homu: {"type":"BuildStarted","head_sha":"6a66aa4a263a3cd7586b723ec589c2f739e233ef","merge_sha":"714bbd995c711cb7242dbf66cb6aa2c3a238cfb7"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1551650198)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1551650198","builder_name":"checks-actions"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #80087) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

Mark-Simulacrum commented:
r=me with conflicts resolved, though if we don't land it today/tomorrow I'd rather wait until early January to avoid breakage around release week (beta promotion is friday) and holidays.

pietroalbini commented:
@bors r=Mark-Simulacrum p=10

bors commented:
:pushpin: Commit 092f535824865f6d8597e9962f6d473021ab7669 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum 092f535824865f6d8597e9962f6d473021ab7669 -->
<!-- homu: {"type":"Approved","sha":"092f535824865f6d8597e9962f6d473021ab7669","approver":"Mark-Simulacrum"} -->

bors commented:
:hourglass: Testing commit 092f535824865f6d8597e9962f6d473021ab7669 with merge fc6fe5e09b2eecbe02427273013268446aa4b79a...
<!-- homu: {"type":"BuildStarted","head_sha":"092f535824865f6d8597e9962f6d473021ab7669","merge_sha":"fc6fe5e09b2eecbe02427273013268446aa4b79a"} -->

rust-log-analyzer commented:
The job **`dist-x86_64-msvc`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/1591369806) [(plain)](https://github.com/rust-lang-ci/rust/commit/fc6fe5e09b2eecbe02427273013268446aa4b79a/checks/1591369806/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 finished in 38.627 seconds
[TIMING] Docs { host: TargetSelection { triple: "x86_64-pc-windows-msvc", file: None } } -- 43.758
[TIMING] RustcDocs { host: TargetSelection { triple: "x86_64-pc-windows-msvc", file: None } } -- 0.000
[TIMING] Mingw { host: TargetSelection { triple: "x86_64-pc-windows-msvc", file: None } } -- 0.000
[TIMING] DebuggerScripts { sysroot: "D:\\a\\rust\\rust\\build\\tmp\\tarball\\x86_64-pc-windows-msvc\\image", host: TargetSelection { triple: "x86_64-pc-windows-msvc", file: None } } -- 0.002
 finished in 41.981 seconds
[TIMING] Rustc { compiler: Compiler { stage: 2, host: TargetSelection { triple: "x86_64-pc-windows-msvc", file: None } } } -- 41.999
Dist rust-std-nightly-x86_64-pc-windows-msvc
 finished in 32.090 seconds
---
 finished in 3.249 seconds
[TIMING] Miri { compiler: Compiler { stage: 1, host: TargetSelection { triple: "x86_64-pc-windows-msvc", file: None } }, target: TargetSelection { triple: "x86_64-pc-windows-msvc", file: None } } -- 3.261
Dist rust-nightly-x86_64-pc-windows-msvc
 finished in 145.862 seconds
thread 'main' panicked at 'could not read dir "D:\\a\\rust\\rust\\build\\tmp/dist\\work\\rustc-nightly-x86_64-pc-windows-msvc\\rustc": Os { code: 3, kind: NotFound, message: "The system cannot find the path specified." }', src\bootstrap\lib.rs:1339:25
failed to run: D:\a\rust\rust\build\bootstrap\debug\bootstrap dist
Build completed unsuccessfully in 1:00:09
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1591653644)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1591653644","builder_name":"checks-actions"} -->

pietroalbini commented:
@bors try

bors commented:
:hourglass: Trying commit f54e9a5837f14371fd884e9841e9829266717e44 with merge ee87086071b8ae244b3f25d4ddc3ca99f7854b29...
<!-- homu: {"type":"TryBuildStarted","head_sha":"f54e9a5837f14371fd884e9841e9829266717e44","merge_sha":"ee87086071b8ae244b3f25d4ddc3ca99f7854b29"} -->

rust-log-analyzer commented:
The job **`dist-x86_64-msvc`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/1594456750) [(plain)](https://github.com/rust-lang-ci/rust/commit/ee87086071b8ae244b3f25d4ddc3ca99f7854b29/checks/1594456750/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
##[endgroup]
##[group]Determining the checkout info
##[endgroup]
##[group]Checking out the ref
[command]"C:\Program Files\Git\bin\git.exe" checkout --progress --force -B try refs/remotes/origin/try
Updating files:  15% (3950/26327)
Updating files:  16% (4213/26327)
Updating files:  17% (4476/26327)
Updating files:  18% (4739/26327)
---
Updating files:  98% (25801/26327)
Updating files:  99% (26064/26327)
Updating files: 100% (26327/26327)
Updating files: 100% (26327/26327), done.
Branch 'try' set up to track remote branch 'try' from 'origin'.
Switched to a new branch 'try'
[command]"C:\Program Files\Git\bin\git.exe" log -1 --format='%H'
'ee87086071b8ae244b3f25d4ddc3ca99f7854b29'
##[group]Run src/ci/scripts/setup-environment.sh
src/ci/scripts/setup-environment.sh
---
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
file:.git/config submodule.library/stdarch.url=https://github.com/rust-lang/stdarch.git
file:.git/config submodule.src/doc/edition-guide.active=true
---
Dist miri-nightly-x86_64-pc-windows-msvc
 finished in 3.813 seconds
Dist rust-nightly-x86_64-pc-windows-msvc
 finished in 158.529 seconds
thread 'main' panicked at 'could not read dir "D:\\a\\rust\\rust\\build\\tmp/dist\\work\\rustc-nightly-x86_64-pc-windows-msvc\\rustc": Os { code: 3, kind: NotFound, message: "The system cannot find the path specified." }', src\bootstrap\lib.rs:1339:25
stack backtrace:
   0: std::panicking::begin_panic_handler
             at /rustc/21dea46d8347c8b4117c5567949703f0fbb51649\/library\std\src\panicking.rs:495
   1: std::panicking::begin_panic_fmt
             at /rustc/21dea46d8347c8b4117c5567949703f0fbb51649\/library\std\src\panicking.rs:437
   2: bootstrap::Build::read_dir
             at .\src\bootstrap\lib.rs:1339
   3: bootstrap::Build::cp_r
             at .\src\bootstrap\lib.rs:1234
   4: bootstrap::dist::{{impl}}::run::{{closure}}
             at .\src\bootstrap\dist.rs:1455
   5: bootstrap::dist::{{impl}}::run
             at .\src\bootstrap\dist.rs:1461
   6: bootstrap::builder::Builder::ensure<bootstrap::dist::Extended>
             at .\src\bootstrap\builder.rs:1487
   7: bootstrap::dist::{{impl}}::make_run
             at .\src\bootstrap\dist.rs:1252
   8: bootstrap::builder::StepDescription::maybe_run
             at .\src\bootstrap\builder.rs:179
   9: bootstrap::builder::StepDescription::run
             at .\src\bootstrap\builder.rs:200
  10: bootstrap::builder::Builder::run_step_descriptions
             at .\src\bootstrap\builder.rs:570
  11: bootstrap::builder::Builder::execute_cli
             at .\src\bootstrap\builder.rs:561
  12: bootstrap::Build::build
             at .\src\bootstrap\lib.rs:510
  13: bootstrap::main
             at .\src\bootstrap\bin\main.rs:30
  14: core::ops::function::FnOnce::call_once<fn(),tuple<>>
             at /rustc/21dea46d8347c8b4117c5567949703f0fbb51649\library\core\src\ops\function.rs:227
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
failed to run: D:\a\rust\rust\build\bootstrap\debug\bootstrap dist
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1594703888)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1594703888","builder_name":"checks-actions"} -->

pietroalbini commented:
@bors try

bors commented:
:hourglass: Trying commit a4a3b47c58d6da764f007d80f114370a6b7f874c with merge 5d9e7a91950efaeaf0413e6027147c18eaec3913...
<!-- homu: {"type":"TryBuildStarted","head_sha":"a4a3b47c58d6da764f007d80f114370a6b7f874c","merge_sha":"5d9e7a91950efaeaf0413e6027147c18eaec3913"} -->

rust-log-analyzer commented:
The job **`dist-x86_64-msvc`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/1594788447) [(plain)](https://github.com/rust-lang-ci/rust/commit/5d9e7a91950efaeaf0413e6027147c18eaec3913/checks/1594788447/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
##[endgroup]
##[group]Determining the checkout info
##[endgroup]
##[group]Checking out the ref
[command]"C:\Program Files\Git\bin\git.exe" checkout --progress --force -B try refs/remotes/origin/try
Updating files:  15% (3950/26327)
Updating files:  16% (4213/26327)
Updating files:  17% (4476/26327)
Updating files:  18% (4739/26327)
---
Updating files:  98% (25801/26327)
Updating files:  99% (26064/26327)
Updating files: 100% (26327/26327)
Updating files: 100% (26327/26327), done.
Branch 'try' set up to track remote branch 'try' from 'origin'.
Switched to a new branch 'try'
[command]"C:\Program Files\Git\bin\git.exe" log -1 --format='%H'
'5d9e7a91950efaeaf0413e6027147c18eaec3913'
##[group]Run src/ci/scripts/setup-environment.sh
src/ci/scripts/setup-environment.sh
---
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
file:.git/config submodule.library/stdarch.url=https://github.com/rust-lang/stdarch.git
file:.git/config submodule.src/doc/edition-guide.active=true
---
Dist miri-nightly-x86_64-pc-windows-msvc
 finished in 3.702 seconds
Dist rust-nightly-x86_64-pc-windows-msvc
 finished in 157.650 seconds
thread 'main' panicked at 'could not read dir "D:\\a\\rust\\rust\\build\\tmp\\tarball\\x86_64-pc-windows-msvc\\rustc-nightly-x86_64-pc-windows-msvc\\rustc": Os { code: 3, kind: NotFound, message: "The system cannot find the path specified." }', src\bootstrap\lib.rs:1339:25
stack backtrace:
   0: std::panicking::begin_panic_handler
             at /rustc/21dea46d8347c8b4117c5567949703f0fbb51649\/library\std\src\panicking.rs:495
   1: std::panicking::begin_panic_fmt
             at /rustc/21dea46d8347c8b4117c5567949703f0fbb51649\/library\std\src\panicking.rs:437
   2: bootstrap::Build::read_dir
             at .\src\bootstrap\lib.rs:1339
   3: bootstrap::Build::cp_r
             at .\src\bootstrap\lib.rs:1234
   4: bootstrap::dist::{{impl}}::run::{{closure}}
             at .\src\bootstrap\dist.rs:1455
   5: bootstrap::dist::{{impl}}::run
             at .\src\bootstrap\dist.rs:1461
   6: bootstrap::builder::Builder::ensure<bootstrap::dist::Extended>
             at .\src\bootstrap\builder.rs:1487
   7: bootstrap::dist::{{impl}}::make_run
             at .\src\bootstrap\dist.rs:1252
   8: bootstrap::builder::StepDescription::maybe_run
             at .\src\bootstrap\builder.rs:179
   9: bootstrap::builder::StepDescription::run
             at .\src\bootstrap\builder.rs:200
  10: bootstrap::builder::Builder::run_step_descriptions
             at .\src\bootstrap\builder.rs:570
  11: bootstrap::builder::Builder::execute_cli
             at .\src\bootstrap\builder.rs:561
  12: bootstrap::Build::build
             at .\src\bootstrap\lib.rs:510
  13: bootstrap::main
             at .\src\bootstrap\bin\main.rs:30
  14: core::ops::function::FnOnce::call_once<fn(),tuple<>>
             at /rustc/21dea46d8347c8b4117c5567949703f0fbb51649\library\core\src\ops\function.rs:227
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
failed to run: D:\a\rust\rust\build\bootstrap\debug\bootstrap dist
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1595051166)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1595051166","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1596466291) [(plain)](https://github.com/rust-lang/rust/commit/77c29c434953ad7e0ab612121941b29553a761a8/checks/1596466291/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling ignore v0.4.16
   Compiling merge_derive v0.1.0
   Compiling merge v0.1.0
   Compiling toml v0.5.7
error[E0382]: use of partially moved value: `self`
   --> src/bootstrap/tarball.rs:220:9
    |
216 |                     .expect("include_target_in_component_name used in a targetless tarball"),
    |                      ----------------------------------------------------------------------- `self.target` partially moved due to this method call
...
220 |         self.run(|this, cmd| {
    |         ^^^^ value used here after partial move
    |
note: this function consumes the receiver `self` by taking ownership of it, which moves `self.target`
    = note: partial move occurs because `self.target` has type `std::option::Option<std::string::String>`, which does not implement the `Copy` trait
error: aborting due to previous error

For more information about this error, try `rustc --explain E0382`.
error: could not compile `bootstrap`
```

</details>

pietroalbini commented:
Made a few changes that should *hopefully* fix Windows (will reboot into it and test it locally in the next few minutes):

* There is now a method to avoid deleting the temporary directory at the end of the tarball generation, which is used by the Windows package generation as part of the `rust` component.
* There are now the `is_preview(true)` and `include_target_in_component_name(true)` methods to properly set the component name of some tarballs (fixing a regression that would've been introduced by this PR).

r? @Mark-Simulacrum

pietroalbini commented:
Ok this PR should work on Windows too (checked locally).

Mark-Simulacrum commented:
@bors r+ rollup=never

bors commented:
:pushpin: Commit a61260353d69121fcd3917614dd4d882d0e1d877 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum a61260353d69121fcd3917614dd4d882d0e1d877 -->
<!-- homu: {"type":"Approved","sha":"a61260353d69121fcd3917614dd4d882d0e1d877","approver":"Mark-Simulacrum"} -->

bors commented:
:hourglass: Testing commit a61260353d69121fcd3917614dd4d882d0e1d877 with merge ae9cf923e2e56aafaf89a5d101a11a4c49a20ff3...
<!-- homu: {"type":"BuildStarted","head_sha":"a61260353d69121fcd3917614dd4d882d0e1d877","merge_sha":"ae9cf923e2e56aafaf89a5d101a11a4c49a20ff3"} -->

rust-log-analyzer commented:
The job **`armhf-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/1601149389) [(plain)](https://github.com/rust-lang-ci/rust/commit/ae9cf923e2e56aafaf89a5d101a11a4c49a20ff3/checks/1601149389/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling toml v0.5.7
error[E0425]: cannot find function `rust_installer` in this scope
    --> src/bootstrap/dist.rs:2060:23
     |
2060 |         let mut cmd = rust_installer(builder);

error: aborting due to previous error

For more information about this error, try `rustc --explain E0425`.
---
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0425]: cannot find function `rust_installer` in this scope
    --> src/bootstrap/dist.rs:2060:23
     |
2060 |         let mut cmd = rust_installer(builder);

error: aborting due to previous error

For more information about this error, try `rustc --explain E0425`.
---
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0425]: cannot find function `rust_installer` in this scope
    --> src/bootstrap/dist.rs:2060:23
     |
2060 |         let mut cmd = rust_installer(builder);

error: aborting due to previous error

For more information about this error, try `rustc --explain E0425`.
---
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0425]: cannot find function `rust_installer` in this scope
    --> src/bootstrap/dist.rs:2060:23
     |
2060 |         let mut cmd = rust_installer(builder);

error: aborting due to previous error

For more information about this error, try `rustc --explain E0425`.
---
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0425]: cannot find function `rust_installer` in this scope
    --> src/bootstrap/dist.rs:2060:23
     |
2060 |         let mut cmd = rust_installer(builder);

error: aborting due to previous error

For more information about this error, try `rustc --explain E0425`.
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1601175465)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1601175465","builder_name":"checks-actions"} -->

pietroalbini commented:
A PR landed with the legacy tarball creation code. Will rebase shortly.

pietroalbini commented:
Rebased on the latest master and added the last commit to convert `reproducible-artifacts` to use `Tarball`.

r? @Mark-Simulacrum 
@rustbot modify labels: -S-waiting-on-author S-waiting-on-review

Mark-Simulacrum commented:
@bors r+

bors commented:
:pushpin: Commit 8736731f70be86b169a0dcc6a1900ccc05d2db85 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum 8736731f70be86b169a0dcc6a1900ccc05d2db85 -->
<!-- homu: {"type":"Approved","sha":"8736731f70be86b169a0dcc6a1900ccc05d2db85","approver":"Mark-Simulacrum"} -->

bors commented:
:hourglass: Testing commit 8736731f70be86b169a0dcc6a1900ccc05d2db85 with merge 7f9c43cf98cfe1c369045399929cb098155b8374...
<!-- homu: {"type":"BuildStarted","head_sha":"8736731f70be86b169a0dcc6a1900ccc05d2db85","merge_sha":"7f9c43cf98cfe1c369045399929cb098155b8374"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1602515660)
Approved by: Mark-Simulacrum
Pushing 7f9c43cf98cfe1c369045399929cb098155b8374 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1602515660"},"merge_sha":"7f9c43cf98cfe1c369045399929cb098155b8374"} -->

semarie commented:
if I didn't mess myself, with this PR, the current nightly tarball extracts inside `image/` directory instead of previously `rustc-nightly-src/` directory.

does such name will stick on release too (instead of `rustc-1.50-src/`) ?

Mark-Simulacrum commented:
Yeah, at least rustc-nightly-src changed. Looking into a fix now.

Mark-Simulacrum commented:
Fix should be up in https://github.com/rust-lang/rust/pull/80397, and seems to work locally.

pietroalbini commented:
Sorry about that!
